pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,337 @@
+		return nil
+	}
+	clusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
+
+	obj, err := c.workspaceLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+	previous := obj
+	obj = obj.DeepCopy()
+
+	if err := c.reconcile(ctx, obj); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous.Status, obj.Status) {
+		oldData, err := json.Marshal(tenancyv1alpha1.Workspace{
+			Status: previous.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal old data for workspace %q|%q/%q: %v", clusterName, namespace, name, err)
+		}
+
+		newData, err := json.Marshal(tenancyv1alpha1.Workspace{
+			ObjectMeta: metav1.ObjectMeta{
+				UID:             previous.UID,
+				ResourceVersion: previous.ResourceVersion,
+			}, // to ensure they appear in the patch as preconditions
+			Status: obj.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal new data for workspace %q|%q/%q: %v", clusterName, namespace, name, err)
+		}
+
+		patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+		if err != nil {
+			return fmt.Errorf("failed to create patch for workspace %q|%q/%q: %v", clusterName, namespace, name, err)
+		}
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().Workspaces().Patch(ctx, obj.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+		return uerr
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, workspace *tenancyv1alpha1.Workspace) error {
+	if workspace.Status.BaseURL == "" {
+		// TODO: something sensible when we have some use for this field
+		workspace.Status.BaseURL = fmt.Sprintf("%s.workspaces.kcp.dev", workspace.Name)
+	}
+	var shouldReSchedule bool
+	if currentShard := workspace.Status.Location.Current; currentShard != "" {
+		// make sure current shard still exists
+		_, err := c.workspaceShardLister.Get(clusters.ToClusterAwareKey(workspace.ClusterName, currentShard))
+		if errors.IsNotFound(err) {
+			klog.Infof("de-scheduling workspace %q from nonexistent shard %q", workspace.Name, currentShard)
+			shouldReSchedule = true
+		} else if err != nil {
+			return err
+		}
+	}
+	if workspace.Status.Location.Current == "" || shouldReSchedule {
+		// find a shard for this workspace
+		shards, err := c.workspaceShardLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		if len(shards) != 0 {
+			target := shards[rand.Intn(len(shards))].Name
+			workspace.Status.Location.Target = target
+			klog.Infof("scheduling workspace %q to %q", workspace.Name, target)
+		}
+	}
+	if workspace.Status.Location.Current != workspace.Status.Location.Target {
+		klog.Infof("moving workspace %q from %q to %q", workspace.Name, workspace.Status.Location.Current, workspace.Status.Location.Target)
```

Comment:
Thinking through at least the log messages: if we have a new unscheduled workspace (current=""), we'll see "scheduling workspace Foo to Bar" from L301, then I'm assuming because current is still empty and we just set target, we'll see "moving workspace Foo from '' to 'Bar'"?

Reply:
Likely below our bar for worrying about it right now? The flow of this controller is 100% guaranteed to change in the short term when we have actual scheduling concepts in

Reply:
Yes, totally fine, thanks